Fix resource targetPath resolution to be relative to output directory (fixes #11381)

gnodet opened the pull request:
This PR fixes a regression in Maven 4 where resource `targetPath` was being resolved relative to the project base directory instead of the output directory (`target/classes` or `target/test-classes`).

## Problem

In Maven 3.x, when a resource has a relative `targetPath`, it's resolved relative to the output directory. For example:

```xml
<resource>
  <directory>${project.basedir}/rest</directory>
  <targetPath>target-dir</targetPath>
  <includes>
    <include>**/*.yml</include>
  </includes>
</resource>
```

In Maven 3.x, files would be copied to `target/classes/target-dir/`, but in Maven 4, they were being copied to `target-dir/` (relative to project root).

## Root Cause

The bug was in the `DefaultSourceRoot` constructor that takes a `Resource` parameter. On line 172, the targetPath was being resolved as:
```java
nonBlank(resource.getTargetPath()).map(baseDir::resolve).orElse(null)
```

This resolves the targetPath relative to the base directory, not the output directory.

## Solution

1. **Created a new constructor** in `DefaultSourceRoot.java` that accepts an `outputDir` parameter and correctly resolves targetPath relative to the output directory
2. **Deprecated the old constructor** to maintain backward compatibility
3. **Updated all call sites** to use the new constructor:
   - `DefaultProjectBuilder.java`
   - `ConnectedResource.java`
   - `MavenProject.java`
4. **Updated unit tests** in `DefaultSourceRootTest.java` to verify the correct behavior
5. **Created an integration test** `MavenITgh11381ResourceTargetPathTest.java` to verify the fix works end-to-end

## Testing

The integration test verifies that:
- Resources with relative `targetPath` are copied to the correct location under the output directory
- The fix works for both main and test scopes

Test results:
```
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```

Fixes #11381

---
Pull Request opened by [Augment Code](https://www.augmentcode.com/) with guidance from the PR author

desruisseaux commented on impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultSourceRoot.java:
```diff
@@ -169,7 +169,7 @@ public DefaultSourceRoot(final Path baseDir, ProjectScope scope, Resource resour
                 resource.getExcludes(),
                 Boolean.parseBoolean(resource.getFiltering()),
-                nonBlank(resource.getTargetPath()).map(baseDir::resolve).orElse(null),
+                nonBlank(resource.getTargetPath()).map(Path::of).orElse(null),
```
> It makes the path truly relative, and later resolved against the correct directory.

Ah, I see. The responsibility of resolving the path is shifted to the `SourceRoot` user instead of the `SourceRoot` constructor. But why not resolving at construction time? It would probably be less code to check for correctness.

gnodet replied:
I agree.  I think that would be desired behavior, but then we need to make it truly relative in the ConnectedResource to not break the existing plugin.
So the behavior would be different between mvn3 and mvn4 api, but that's fine.

gnodet commented:
## Implementation Details

I've implemented a fix for this issue that ensures `SourceRoot.targetPath()` and `ConnectedResource` behave correctly according to the Maven 4 API while preserving Maven 3 compatibility.

### Root Cause

The regression occurred because `DefaultSourceRoot.fromModel()` was storing `targetPath` as an **absolute path** (resolved against `baseDir` and `outputDir`), when it should have been stored as a **relative path** according to the Maven 4 API design.

### Maven 4 API Design

The Maven 4 API defines two methods for `SourceRoot`:

1. **`targetPath()`** - Returns an `Optional<Path>` containing the explicit target path. According to the javadoc and the presence of the resolution method below, this should be a **relative path** (or absolute if explicitly specified as absolute).

2. **`targetPath(Project)`** - Returns the resolved absolute path by taking the relative `targetPath()` and resolving it against the project's output directory:
   ```java
   Path base = project.getOutputDirectory(scope());
   return targetPath.map(base::resolve).orElse(base);
   ```

This two-method design clearly indicates that `targetPath()` should return a **relative** path that gets resolved by `targetPath(Project)`.

### Implementation

**1. Fixed `DefaultSourceRoot.fromModel()`** (impl/maven-impl)
```java
// Before (incorrect):
nonBlank(source.getTargetPath())
    .map((targetPath) -> baseDir.resolve(outputDir.apply(scope)).resolve(targetPath))
    .orElse(null)

// After (correct):
nonBlank(source.getTargetPath()).map(Path::of).orElse(null)
```

Now `targetPath()` returns a **relative path** as intended, and `targetPath(Project)` handles the resolution.

**2. Simplified `ConnectedResource.computeRelativeTargetPath()`** (impl/maven-core)

Since `SourceRoot.targetPath()` now always returns a relative path, we can simply return it as-is:
```java
return sourceRoot.targetPath().map(Path::toString).orElse(null);
```

This removed ~40 lines of complex logic that was trying to convert absolute paths back to relative paths.

### Maven 3 Compatibility

In Maven 3, `Resource.getTargetPath()` was always **relative to the output directory**. This behavior is preserved:

- `SourceRoot.targetPath()` stores the path as relative (e.g., `"custom-dir"`)
- `ConnectedResource` converts it back to a String for the Maven 3 `Resource` API
- Resources are copied to `target/classes/custom-dir` (Maven 3 behavior) ✅

### Example

With `<targetPath>custom-dir</targetPath>` in a resource configuration:

- **Maven 3**: Resources copied to `target/classes/custom-dir` ✅
- **Maven 4 (before fix)**: Resources copied to `project-root/custom-dir` ❌
- **Maven 4 (after fix)**: Resources copied to `target/classes/custom-dir` ✅

### Testing

All existing tests pass, including:
- `DefaultSourceRootTest` - Updated to expect relative paths from `targetPath()`
- `ResourceIncludeTest` - Validates the Maven 3 compatibility layer
- Integration test `MavenITgh11381ResourceTargetPathTest` should now pass

The fix is minimal, focused, and aligns with the Maven 4 API design while maintaining backward compatibility.

gnodet commented:
## Latest Update: Enhanced Javadoc for API Clarity

I've just pushed an update that significantly improves the documentation to clarify the Maven 4 API semantics while making the Maven 3 backward compatibility guarantees explicit.

### What Changed

Enhanced javadoc for three key methods:

1. **`SourceRoot.targetPath()`** - Now clearly documents:
   - Relative paths are resolved relative to the output directory (not project basedir)
   - Absolute paths are used as-is
   - Empty/null means files go to output directory root
   - Explicitly states Maven 3 compatibility is maintained

2. **`SourceRoot.targetPath(Project)`** - Now includes:
   - Step-by-step resolution algorithm
   - Concrete examples with actual paths
   - Cross-references to related methods

3. **`Project.getOutputDirectory(ProjectScope)`** - Now clarifies:
   - Returns directory for both compiled classes AND resources
   - Its role in SourceRoot targetPath resolution
   - Maven 3 compatibility semantics

4. **`DefaultSourceRoot` constructor from Resource** - Now documents:
   - targetPath is stored as-is (not resolved against basedir)
   - baseDir parameter is only used for source directory resolution
   - How this preserves Maven 3.x behavior

### Why This Matters

The enhanced documentation makes it crystal clear that:

- **For Maven 4 API users**: Use `targetPath(Project)` to get the fully resolved absolute path
- **For Maven 3 compatibility**: The `targetPath()` method returns the relative path that plugins expect
- **For implementers**: The separation of concerns between storage and resolution is now explicit

### Implementation Verification

The documentation now matches the actual implementation:

✅ `DefaultSourceRoot` stores `targetPath` as-is (relative or absolute)

✅ `ConnectedResource` extracts `targetPath` as a relative path string

✅ maven-resources-plugin receives the relative path and resolves it (Maven 3 behavior)

✅ Maven 4 API consumers can use `targetPath(Project)` for absolute paths

See the detailed explanation in [this comment on #11381](https://github.com/apache/maven/issues/11381#issuecomment-3502314999) for the complete design rationale.
